tests/drivers/uart/uart_async_api/src/test_uart_async.c

```diff
@@ -834,8 +834,8 @@ static void test_chained_write_callback(const struct device *dev,
 	switch (evt->type) {
 	case UART_TX_DONE:
 		if (chained_write_next_buf) {
-			uart_tx(dev, chained_write_tx_bufs[1], 10, 100 * USEC_PER_MSEC);
 			chained_write_next_buf = false;
+			uart_tx(dev, chained_write_tx_bufs[1], 10, 100 * USEC_PER_MSEC);
```

Comment:
is this order swap really needed? it is checked in next `UART_TX_DONE` interrupt.

Reply:
For the case I described this is the sequence:
- uart_tx enabled tx interrupt
- interrupt fires due to 32 byte FIFO being empty
- code in interrupt fills 10 bytes and notifies event UART_TX_DONE (since there is no way to do it when data actually leaves). At this point FIFO can receive 22 more bytes.
- event code checks `chained_write_nex_buf` and see that second part should be send
- uart_tx adds additional 10 bytes to FIFO (around 12 bytes is still free)
- interrupt for FIFO not full fires again and new event UART_TX_DONE is sent
- at this point `chained_write_nex_buf` is still false so extra 10 bytes are send with uart_tx (fifo is still not full)

test_cained_write_callback is called recursively 3 times

Reply:
I think I recalled it incorrectly:
Actually problem is that we only have TX fifo not full interrupt
So if we put 10 bytes in fifo we have to disable interrupt because there is nothing more to send.
uart_tx fills fifo (with interrupts or not) and then it executes callback with event UART_TX_DONE which in turn calls uart_tx again and cycle continues and here is the recursion that I saw.

Reply:
I see, so event handler is not called from the interrupt handler but directly from `uart_tx`.